crates/emitter/src/control_structures.rs

```diff
@@ -0,0 +1,220 @@
+#[must_use]
+pub struct JumpPatchEmitter {
+    offsets: Vec<BytecodeOffset>,
+    depth: usize,
+}
+
+impl JumpPatchEmitter {
+    pub fn patch_merge(self, emitter: &mut AstEmitter) {
+        emitter.emit.patch_and_emit_jump_target(self.offsets);
+
+        // If the previous opcode fall-through, it should have the same stack
+        // depth.
+        debug_assert!(emitter.emit.stack_depth() == self.depth);
+    }
+
+    pub fn patch_not_merge(self, emitter: &mut AstEmitter) {
+        emitter.emit.patch_and_emit_jump_target(self.offsets);
+        // If the previous opcode doesn't fall-through, overwrite the stack
+        // depth.
+        emitter.emit.set_stack_depth(self.depth);
+    }
+}
+
+// Struct for emitting bytecode for forward jump.
+#[derive(Debug)]
+pub struct ForwardJumpEmitter {
+    pub jump: JumpKind,
+}
+impl Jump for ForwardJumpEmitter {
+    fn jump_kind(&mut self) -> &JumpKind {
+        &self.jump
+    }
+}
+
+impl ForwardJumpEmitter {
+    pub fn emit(&mut self, emitter: &mut AstEmitter) -> JumpPatchEmitter {
+        let offsets = vec![emitter.emit.bytecode_offset()];
+        self.emit_jump(emitter);
+        let depth = emitter.emit.stack_depth();
+
+        // The JITs rely on a jump target being emitted after the
+        // conditional jump
+        if self.should_fallthrough() {
+            emitter.emit.jump_target();
+        }
+        JumpPatchEmitter { offsets, depth }
+    }
+}
+
+// Compared to C++ impl, this uses explicit stack struct,
+// given Rust cannot store a reference of stack-allocated object into
+// another object that has longer-lifetime.
+//
+// Some methods are implemented here, instead of each EmitterScope.
+pub struct LoopStack {
+    loop_stack: Vec<LoopControl>,
+}
+
+impl LoopStack {
+    pub fn new() -> Self {
+        Self {
+            loop_stack: Vec::new(),
+        }
+    }
+
+    pub fn open_loop(&mut self, emit: &mut InstructionWriter) {
+        let depth = (self.loop_stack.len() + 1) as u8;
+        let new_loop = LoopControl::new(emit, depth);
+        self.loop_stack.push(new_loop);
+    }
+
+    pub fn register_break(&mut self, offset: BytecodeOffset) {
+        let innermost = self.innermost();
+        innermost.register_break(offset);
+    }
+
+    pub fn close_loop(&mut self, emit: &mut InstructionWriter) {
+        let innermost = self
+            .loop_stack
+            .pop()
+            .expect("There should be at least one loop");
+        innermost.emit_end_target(emit);
+    }
+
+    pub fn innermost(&mut self) -> &mut LoopControl {
+        self.loop_stack
+            .last_mut()
+            .expect("There should be at least one loop")
+    }
+}
+
+pub struct LoopControl {
+    pub breaks: Vec<BytecodeOffset>,
+    pub continues: Vec<BytecodeOffset>,
+    head: BytecodeOffset,
+}
+
+impl LoopControl {
+    pub fn new(emit: &mut InstructionWriter, depth: u8) -> Self {
+        let mut offset = emit.bytecode_offset();
+        if offset.offset == 0 {
```

Comment:
can you add the reference to the comment in `LoopControl::emitLoopHead` ?
so that it's clear what this is doing.